exporter/signalfxexporter/dimensions/dimclient.go

```diff
@@ -0,0 +1,340 @@
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// DimensionClient sends updates to dimensions to the SignalFx API
+// This is a port of https://github.com/signalfx/signalfx-agent/blob/master/pkg/core/writer/dimensions/client.go
+// with the only major difference being deduplication of dimension
+// updates are currently not done by this port.
+type DimensionClient struct {
+	sync.RWMutex
+	ctx           context.Context
+	Token         string
+	APIURL        *url.URL
+	client        *http.Client
+	requestSender *ReqSender
+	// How long to wait for property updates to be sent once they are
+	// generated.  Any duplicate updates to the same dimension within this time
+	// frame will result in the latest property set being sent.  This helps
+	// prevent spurious updates that get immediately overwritten by very flappy
+	// property generation.
+	sendDelay time.Duration
+	// Set of dims that have been queued up for sending.  Use map to quickly
+	// look up in case we need to replace due to flappy prop generation.
+	delayedSet map[DimensionKey]*DimensionUpdate
+	// Queue of dimensions to update.  The ordering should never change once
+	// put in the queue so no need for heap/priority queue.
+	delayedQueue chan *queuedDimension
+	// For easier unit testing
+	now func() time.Time
+
+	// TODO: Look into collecting these metrics and other traces via obsreport
+	DimensionsCurrentlyDelayed int64
+	TotalDimensionsDropped     int64
+	// The number of dimension updates that happened to the same dimension
+	// within sendDelay.
+	TotalFlappyUpdates           int64
+	TotalClientError4xxResponses int64
+	TotalRetriedUpdates          int64
+	TotalInvalidDimensions       int64
+	TotalSuccessfulUpdates       int64
+	logUpdates                   bool
+	logger                       *zap.Logger
+}
+
+type queuedDimension struct {
+	*DimensionUpdate
+	TimeToSend time.Time
+}
+
+type DimensionClientOptions struct {
+	Token      string
+	APIURL     *url.URL
+	LogUpdates bool
+	Logger     *zap.Logger
+	SendDelay  int
+}
+
+// NewDimensionClient returns a new client
+func NewDimensionClient(ctx context.Context, options DimensionClientOptions) *DimensionClient {
+
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   5 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:        20,
+			MaxIdleConnsPerHost: 20,
+			IdleConnTimeout:     30 * time.Second,
+			TLSHandshakeTimeout: 10 * time.Second,
+		},
+	}
+	sender := NewReqSender(ctx, client, 20, map[string]string{"client": "dimension"})
+
+	return &DimensionClient{
+		ctx:           ctx,
+		Token:         options.Token,
+		APIURL:        options.APIURL,
+		sendDelay:     time.Duration(options.SendDelay) * time.Second,
+		delayedSet:    make(map[DimensionKey]*DimensionUpdate),
+		delayedQueue:  make(chan *queuedDimension, 10000),
```

Comment:
should it be a constant or part of the `DimensionClientOptions`? Especially since it's being referenced as `propertiesMaxBuffered` and `conf.PropertiesMaxBuffered` in other places